### PR TITLE
KAFKA-17880: Move integration test from streams module to streams/integration-tests module

### PR DESCRIPTION
**Pull request by FrankYang0529**
Streams module is public so it must be under java 11, but the integration tests relays on server code which is using JDK 17. Hence, we have to move all integration tests to another module

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -2838,6 +2829,55 @@ project(':streams:streams-scala') {
...
+    archivesName = "kafka-streams-integration-tests"
+  }
+
+  dependencies {
+    api project(':streams')
+    api project(':clients')
+
+    implementation libs.slf4jApi
+
+    // testCompileOnly prevents streams from exporting a dependency on test-utils, which would cause a dependency cycle
```
Comment (chia7712) [inline build.gradle]: Is there a circular dependency?

File: streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/AdjustStreamThreadCountTest.java [hunk 2]
```diff
@@ -410,7 +410,8 @@ public void shouldResizeCacheAfterThreadRemovalTimesOut() throws InterruptedExce
             addStreamStateChangeListener(kafkaStreams);
             startStreamsAndWaitForRunning(kafkaStreams);
 
-            try (final LogCaptureAppender appender = LogCaptureAppender.createAndRegister(KafkaStreams.class)) {
+            try (final LogCaptureAppender appender = LogCaptureAppender.createAndRegister()) {
+                appender.setClassLogger(KafkaStreams.class, Level.INFO);
```
Comment (chia7712) [inline streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/AdjustStreamThreadCountTest.java]: Could you please add `log4j.properties` to this module?

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for this patch.

Could you please move `EmbeddedKafkaCluster` from test to production? That can be shared to other streams modules which want to have their integration tests.

File: build.gradle [hunk 3]
```diff
@@ -2803,6 +2793,7 @@ project(':streams:streams-scala') {
     testImplementation project(':server-common').sourceSets.test.output
     testImplementation project(':streams').sourceSets.test.output
     testImplementation project(':clients').sourceSets.test.output
+    testImplementation project(':streams:integration-tests').sourceSets.test.output
```
Comment (chia7712) [inline build.gradle]: `streams` module can't depend on `integration-tests` as all integration tests should be moved from `stream` to `integration-tests` already.

Comment (FrankYang0529) [reply build.gradle]: Updated it. The failed case will be handled in https://github.com/apache/kafka/pull/17645.

File: build.gradle [hunk 4]
```diff
@@ -2848,6 +2833,55 @@ project(':streams:streams-scala') {
...
 
+project(':streams:integration-tests') {
+  apply plugin: 'scala'
+
+  base {
+    archivesName = "kafka-streams-integration-tests"
+  }
+
+  dependencies {
+    api project(':streams')
```
Comment (chia7712) [inline build.gradle]: we don't have production code, so could you please use `testImplementation`

File: build.gradle [hunk 5]
```diff
@@ -2848,6 +2833,55 @@ project(':streams:streams-scala') {
...
+    api project(':clients')
```
Comment (chia7712) [inline build.gradle]: ditto

File: build.gradle [hunk 6]
```diff
@@ -2848,6 +2833,55 @@ project(':streams:streams-scala') {
...
+
+    implementation libs.slf4jApi
```
Comment (chia7712) [inline build.gradle]: ditto

File: build.gradle [hunk 7]
```diff
@@ -2848,6 +2833,55 @@ project(':streams:streams-scala') {
...
+      exclude('kafka-streams*')
+    }
+    into "$buildDir/dependant-libs-${versions.scala}"
+    duplicatesStrategy 'exclude'
+  }
+
+  jar {
+    dependsOn 'copyDependantLibs'
+  }
+
```
Comment (chia7712) [inline build.gradle]: this module have both java and scala code now, so please use joint compilation.

```
  sourceSets {
    // Set java/scala source folders in the `scala` block to enable joint compilation
    main {
      java {
        srcDirs = []
      }
      scala {
        srcDirs = []
      }
    }
    test {
      java {
        srcDirs = []
      }
      scala {
        srcDirs = ["src/test/java", "src/test/scala"]
      }
    }
  }
```

File: build.gradle [hunk 8]
```diff
@@ -2848,6 +2833,55 @@ project(':streams:streams-scala') {
...
+    testImplementation libs.junitJupiter
+    testImplementation libs.junitPlatformSuiteEngine // supports suite test
+    testImplementation libs.mockitoCore
+    testImplementation libs.reload4j
+    testImplementation project(':streams:test-utils')
+
+    testRuntimeOnly runtimeTestLibs
+  }
+
+  tasks.create(name: "copyDependantLibs", type: Copy) {
```
Comment (chia7712) [inline build.gradle]: we don't need to expose the runtime jars for this test-only module

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for your patch

**Comment by FrankYang0529** [comment]
Hi @chia7712, thanks for the great suggestion. I addressed all comments.

File: build.gradle [hunk 9]
```diff
@@ -2848,6 +2833,62 @@ project(':streams:streams-scala') {
...
+    testImplementation libs.junitPlatformSuiteEngine // supports suite test
+    testImplementation libs.mockitoCore
+    testImplementation libs.reload4j
+    testImplementation libs.slf4jApi
+    testImplementation project(':streams:test-utils')
+
+    testRuntimeOnly runtimeTestLibs
+  }
+
+  jar {
```
Comment (chia7712) [inline build.gradle]: remove this as there is no `copyDependantLibs`

File: build.gradle [hunk 10]
```diff
@@ -2803,18 +2793,13 @@ project(':streams:streams-scala') {
     api project(':streams')
 
     api libs.scalaLibrary
-    testImplementation project(':group-coordinator')
-    testImplementation project(':core')
     testImplementation project(':test-common')
```
Comment (chia7712) [inline build.gradle]: remove this unused dependency

**Comment by FrankYang0529** [comment]
Hi @chia7712, thanks for review. I addressed all comments.

**Review by chia7712 — APPROVED** [review]

**Pull request merged**